File: frontend-sdk/packages/skies-react/src/page.ts
// Page<T> — the spine's structural match for the backend's canonical page (Skies.Framework.Abstractions' Page<T>,
// pinned in the OpenAPI document: four members, required, plainly numeric). Structural on purpose, like
// QueryLike: the spine never imports the generated client, so any generated response carrying this shape —
// whatever slice, whatever app — is recognized as a page and the pager hooks compose with it. The numbers
// are the server's EFFECTIVE values after clamping; the contract never reports a page it did not serve.

export interface Page<T> {
  /** The items of this page, in the slice's order — at most `pageSize` of them. */
  items: readonly T[];
  /** The size of the whole filtered set, not of this page — what "21–40 de 87" derives its 87 from. */
  totalCount: number;
  /** The 1-based number of this page, as actually served. */
  pageNumber: number;
  /** The page size actually served. */
  pageSize: number;
}

/** The render facts of a page — see {@link toPageInfo}. */
export interface PageInfo {
  /** Total pages — at least 1 (an empty set is one empty page, so a clamp against it stays sane). */
  pageCount: number;
  /** 1-based index of the first item shown on this page; 0 when the page is empty. */
  from: number;
  /** 1-based index of the last item shown on this page; 0 when the page is empty. */
  to: number;
  totalCount: number;
  hasPrev: boolean;
  hasNext: boolean;
}

/**
 * Derive what a View needs to render a numbered pager — pure, so the ViewModel computes it from the latest
 * `Page` and the View renders "{from}–{to} de {totalCount}" without arithmetic. `undefined` propagates (no
 * page yet → no info), which composes with the clamp story: `pager.next(info?.pageCount)` clamps once the
 * total is known and roams free while it is not.
 */
export function toPageInfo(page: Page<unknown>): PageInfo;
export function toPageInfo(page: Page<unknown> | undefined): PageInfo | undefined;
export function toPageInfo(page: Page<unknown> | undefined): PageInfo | undefined {
  if (page === undefined) return undefined;
  const pageSize = Math.max(1, page.pageSize);
  const pageCount = Math.max(1, Math.ceil(page.totalCount / pageSize));
  const from = page.items.length === 0 ? 0 : (page.pageNumber - 1) * pageSize + 1;
  const to = page.items.length === 0 ? 0 : from + page.items.length - 1;
  return {
    pageCount,
    from,
    to,
    totalCount: page.totalCount,
    hasPrev: page.pageNumber > 1,
    hasNext: page.pageNumber < pageCount,
  };
}

File: flutter-sdk/packages/skies_flutter/lib/src/page.dart
/// The structural page shape shared with the Skies backend contract.
final class Page<T> {
  /// Creates one server-authoritative page.
  Page({
    required Iterable<T> items,
    required this.totalCount,
    required this.pageNumber,
    required this.pageSize,
  }) : items = List<T>.unmodifiable(items);

  /// The items returned for this page.
  final List<T> items;

  /// The total filtered item count across every page.
  final int totalCount;

  /// The one-based page number actually served.
  final int pageNumber;

  /// The page size actually served.
  final int pageSize;
}

/// Render-ready facts derived from one [Page].
final class PageInfo {
  /// Creates pagination facts.
  const PageInfo({
    required this.pageCount,
    required this.from,
    required this.to,
    required this.totalCount,
    required this.hasPrevious,
    required this.hasNext,
  });

  /// The total number of pages, never below one.
  final int pageCount;

  /// The one-based first item position, or zero for an empty result.
  final int from;

  /// The one-based last item position, or zero for an empty result.
  final int to;

  /// The total filtered item count.
  final int totalCount;

  /// Whether a previous page exists.
  final bool hasPrevious;

  /// Whether a following page exists.
  final bool hasNext;
}

/// Derives render facts without leaking pagination arithmetic into a View.
PageInfo? toPageInfo(Page<Object?>? page) {
  if (page == null) return null;
  final pageSize = page.pageSize < 1 ? 1 : page.pageSize;
  final pageCount = (page.totalCount / pageSize).ceil().clamp(1, 1 << 31);
  final from = page.items.isEmpty ? 0 : (page.pageNumber - 1) * pageSize + 1;
  final to = page.items.isEmpty ? 0 : from + page.items.length - 1;
  return PageInfo(
    pageCount: pageCount,
    from: from,
    to: to,
    totalCount: page.totalCount,
    hasPrevious: page.pageNumber > 1,
    hasNext: page.pageNumber < pageCount,
  );
}
